### PR TITLE
Updates to AIChat UI

### PR DESCRIPTION
**Pull request by nullhook**
Resolves https://github.com/brave/brave-browser/issues/30264
Resolves https://github.com/brave/brave-browser/issues/31310

This PR adjusts the current UI to match the designs on Figma and adds support for displaying a favicon and site title for each tab's content. 

The PR also modifies the UI for auto suggestion components.

image
image
image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc [hunk 1]
```diff
@@ -183,6 +219,39 @@ void AIChatUIPageHandler::OnTabStripModelChanged(
...
+      [](GetFaviconImageDataCallback callback,
+         const favicon_base::FaviconRawBitmapResult& result) {
+        if (!result.is_valid()) {
+          std::move(callback).Run(absl::nullopt);
+          return;
+        }
+
+        scoped_refptr<base::RefCountedMemory> bytes = result.bitmap_data;
+        std::vector<uint8_t> buffer(bytes->size());
+        memcpy(buffer.data(), bytes->front_as<uint8_t>(), bytes->size());
```
[the remark this answers is not in the document]
Comment (fmarier) [reply browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:254]: Isn't there a way to initialize the vector with the right values right away?

Something like:
```
vector<uint8_t> buffer(bytes->front_as<unit8_t>(), bytes->end_as<uint8_t>());
```

[the remark this answers is not in the document]
Comment (nullhook) [reply browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:254]: I don't think there is end_as but you can do this:

```c++
std::vector<uint8_t> buffer(bytes->front_as<uint8_t>(), bytes->front_as<uint8_t>() + bytes->size());
```

[the remark this answers is not in the document]
Comment (nullhook) [reply browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc]: changed

File: browser/ui/webui/ai_chat/ai_chat_ui.cc [hunk 2]
```diff
@@ -17,7 +17,9 @@
 #include "brave/components/l10n/common/localization_util.h"
 #include "chrome/browser/profiles/profile.h"
 #include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/webui/favicon_source.h"
 #include "chrome/browser/ui/webui/webui_util.h"
+#include "components/favicon_base/favicon_url_parser.h"
```
Comment (petemill) [inline browser/ui/webui/ai_chat/ai_chat_ui.cc:22]: added by mistake?

Comment (nullhook) [reply browser/ui/webui/ai_chat/ai_chat_ui.cc:22]: yes, along with `#include "services/network/public/mojom/content_security_policy.mojom.h"` I'm not sure if that header is needed

File: browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc [hunk 3]
```diff
@@ -131,6 +141,17 @@ void AIChatUIPageHandler::SetAutoGenerateQuestions(bool value) {
...
 }
 
+void AIChatUIPageHandler::GetSiteInfo(GetSiteInfoCallback callback) {
+  mojom::SiteInfo site_info;
+
+  if (active_chat_tab_helper_) {
+    site_info.title =
+        base::UTF16ToUTF8(active_chat_tab_helper_->web_contents()->GetTitle());
+  }
+
```
[the remark this answers is not in the document]
Comment (nullhook) [reply browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc]: created a private function

File: browser/ui/webui/ai_chat/ai_chat_ui.cc (hunk 2 above)
Comment (nullhook) [reply browser/ui/webui/ai_chat/ai_chat_ui.cc]: removed

File: components/ai_chat/resources/page/components/site_title/index.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import * as React from 'react'
+
+import styles from './style.module.scss'
+import { SiteInfo } from '../../api/page_handler'
+
+interface SiteTitleProps {
+  siteInfo?: SiteInfo
+  favIconUrl: string | null
```
Comment (petemill) [inline components/ai_chat/resources/page/components/site_title/index.tsx:13]: why not optional string?

File: components/ai_chat/resources/page/components/site_title/style.module.scss [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+.imgBox {
+  width: 16px;
+  height: 16px;
+
+  img {
+    width: 100%;
+    height: 100%;
+    object-fit: contain;
+  }
+}
```
Comment (petemill) [inline components/ai_chat/resources/page/components/site_title/style.module.scss:46]: why does the image need to be wrapped in 2 containers? Can't the width and height be on the image?

File: components/ai_chat/resources/page/components/site_title/style.module.scss [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+.box {
+  display: flex;
+  flex-direction: row;
+  justify-content: center;
+  gap: 16px;
+  padding: 14px 0;
+}
+
+.titleBox {
+  width: 100%;
```
Comment (petemill) [inline components/ai_chat/resources/page/components/site_title/style.module.scss:15]: why is this needed? Isn't `width: 100%` the default?

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss [hunk 7]
```diff
@@ -4,35 +4,28 @@
 // you can obtain one at https://mozilla.org/MPL/2.0/.
 
 .list {
-  padding: 16px;
+  // padding: 16px;
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/style.module.scss:7]: remove

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss [hunk 8]
```diff
@@ -4,35 +4,28 @@
...
+  gap: 16px;
+  font: var(--leo-font-secondary-default-regular);
+  color: var(--leo-color-text-primary);
+  padding: 16px;
+}
 
 .turnAI {
-  @extend .textBase;
-  padding: 16px;
+  @extend .turnBase;
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/style.module.scss:20]: IMO let's try not to use sass features, and stick to css. Most styles in brave-core are styled-components. I think we can allow css in lieu of a decision to go forward in a single direction. But adding sass-specific features I think adds yet another framework which not everyone knows the nuances of. OK to keep .scss ending so we get compile-type support for variables which we may start using from the Design System instead of css variables so that we get compiler errors when they change

**Review by petemill — CHANGES_REQUESTED** [review]
Looking good, but some minor feedback. It's looking strange on internal pages don't you think? IMO let's hide the favicon and title from those, in this PR if possible. Perhaps we should only show the favicon and title if the Tab Helper has page content available, otherwise the user may think the conversation is about the page when it isn't since there is no content available.

File: components/ai_chat/resources/page/components/site_title/index.tsx (hunk 4 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/site_title/index.tsx]: made it optional

File: components/ai_chat/resources/page/components/site_title/style.module.scss (hunk 5 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/site_title/style.module.scss]: removed extra container

File: components/ai_chat/resources/page/components/site_title/style.module.scss (hunk 6 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/site_title/style.module.scss]: was a left over probably but i have now added line clamping

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss (hunk 7 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/style.module.scss]: done

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss (hunk 8 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/style.module.scss]: makes sense. done.

File: components/ai_chat/common/mojom/ai_chat.mojom [hunk 9]
```diff
@@ -5,6 +5,8 @@
 
 module ai_chat.mojom;
 
+import "url/mojom/url.mojom";
```
Comment (petemill) [inline components/ai_chat/common/mojom/ai_chat.mojom:8]: this seems like a mistake

File: components/ai_chat/resources/page/state/hooks.ts [hunk 10]
```diff
@@ -4,43 +4,67 @@
...
+
 export function useConversationHistory() {
   const [conversationHistory, setConversationHistory] = React.useState<ConversationTurn[]>([])
   const [isGenerating, setIsGenerating] = React.useState(false)
   const [suggestedQuestions, setSuggestedQuestions] = React.useState<string[]>([])
   const [canGenerateQuestions, setCanGenerateQuestions] = React.useState(false)
-  const [userAllowsAutoGenerating, setUserAllowsAutoGeneratingState] = React.useState(false)
+  const [userAutoGeneratePref, setUserAutoGeneratePref] = React.useState<AutoGenerateQuestionsPref | undefined>(undefined)
+  const [siteInfo, setSiteInfo] = React.useState<SiteInfo | null>(null)
+  const [favIconUrl, setFavIconUrl] = React.useState<string | undefined>(undefined)
```
Comment (petemill) [inline components/ai_chat/resources/page/state/hooks.ts:24]: nit: do these all have to be `| undefined` or `| null`? I believe this is ok to do:
```
const [userAutoGeneratePref, setUserAutoGeneratePref] = React.useState<AutoGenerateQuestionsPref>()
```
It will be `undefined` by default, right?

**Review by petemill — CHANGES_REQUESTED** [review]
Couple issues I can see when running it:
- Page title should be vertically centered, perhaps only if the page title is 1 line long, and top aligned to the favicon when it's longer. I think there's a way to do that in css, can't remember. Perhaps cut off when longer than 2 or 3 lines and have it always vertically centered. https://www.figma.com/file/h7fKSJF2P556bZWfyLTfwD?node-id=849:25297&mode=design#487875663
- Spacing between "Suggested follow-ups" and the buttons underneath is too big. In figma looks like it's 14px or 16px but here it looks 22px+.
- Similarly, spacing between last conversation entry and "Suggested follow-ups" is too small - looks like in figma it should be around 18px-20px from the bottom of the conversation entry text to the top of the "Suggested follow-ups" text.
- Spacing between human conversation entry and "Suggested follow-ups" is non-existant (when assistant conversation entry response hasn't been received yet). Space underneath whole conversation should always be the same, i.e. 18-20px from previous point.
- On "Enable suggested questions?", the vertical alignment of the icon to the title is off - the icon is too high compared to figma. 
image

File: components/ai_chat/common/mojom/ai_chat.mojom (hunk 9 above)
Comment (nullhook) [reply components/ai_chat/common/mojom/ai_chat.mojom:8]: was a left over

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss [hunk 11]
```diff
@@ -4,45 +4,102 @@
...
-  p:first-child {
-    background-color: var(--leo-color-container-highlight);
-    padding: 16px;
-    border-radius: 16px 16px 2px 16px;
-  }
+  display: flex;
+  align-items: flex-start;
+  gap: 16px;
+  font: var(--leo-font-secondary-default-regular);
+  color: var(--leo-color-text-primary);
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/style.module.scss:26]: nit: Instead of repeating these you could add a selector for `.turnHuman, .turnAI`, or add a class name for `.turn`

File: components/ai_chat/resources/page/components/conversation_list/index.tsx [hunk 12]
```diff
@@ -39,45 +35,58 @@ function ConversationList (props: ConversationListProps) {
...
+
+        const avatarStyles = classnames({
+          [styles.avatarAI]: isAIAssistant,
+          [styles.avatarHuman]: isHuman
+        })
+
         return (
-          <div key={id} ref={elementRef} className={turnClass}>
-            <p>
+          <div key={id} ref={elementRef} className={turnClass} data-turn={isHuman ? 'human' : 'assistant'}>
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/index.tsx:61]: Why do we need both the class name which is `turnAI|turnHuman` _and_ the data attribute which is `human|assistant`. It made it difficult to see in the css that they are the same element.

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss [hunk 13]
```diff
@@ -4,45 +4,102 @@
...
+  height: 32px;
+  border-radius: 100px;
+  flex-shrink: 0;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  background: linear-gradient(314.42deg, #FA7250 8.49%, #FF1893 43.72%, #A78AFF 99.51%);
+  color: white;
+}
+
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/style.module.scss:92]: Also a lot of duplication here, seems like we can combine the rules

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss [hunk 14]
```diff
@@ -4,45 +4,102 @@
...
   width: 8px;
   height: 12px;
   background-color: currentColor;
 }
 
-.suggestedQuestions {
+.suggestedQuestionsBox {
+  padding: 16px 16px 16px 64px;
+
+  // TODO(nullhook): Make this change in dev leo
```
Comment (petemill) [inline components/ai_chat/resources/page/components/conversation_list/style.module.scss:44]: Remove this TODO right? We talked about it belonging here.

File: components/ai_chat/resources/page/components/input_box/style.module.scss [hunk 15]
```diff
@@ -3,41 +3,37 @@
...
 
-.container {
-  padding: 16px;
-  border-top: 1px solid var(--leo-color-divider-subtle);
-  background-color: var(--leo-color-container-background);
-}
-
 .form {
-  background-color: var(--leo-color-container-highlight);
+  background-color: transparent;
```
Comment (petemill) [inline components/ai_chat/resources/page/components/input_box/style.module.scss:7]: Isn't `transparent` the default?

File: components/ai_chat/resources/page/components/prompt_auto_suggestion/index.tsx [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+interface PromptAutoSuggestionProps {
+  onEnable?: () => void
+  onDismiss?: () => void
+}
+
+function PromptAutoSuggestion (props: PromptAutoSuggestionProps) {
+  return (
+    <div className={styles.box}>
+      <i><Icon name="product-brave-ai" /></i>
```
Comment (petemill) [inline components/ai_chat/resources/page/components/prompt_auto_suggestion/index.tsx:20]: nit: not sure the `<i>` is neccessary, the className and the style rules can go on the `<Icon>` directly

**Review by petemill — APPROVED** [review]
Works nicely. Left some nits that would be good to clean up before merge.

File: components/ai_chat/resources/page/components/prompt_auto_suggestion/index.tsx (hunk 16 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/prompt_auto_suggestion/index.tsx]: removed

File: components/ai_chat/resources/page/components/input_box/style.module.scss (hunk 15 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/input_box/style.module.scss]: removed

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss (hunk 14 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/style.module.scss]: removed

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss (hunk 13 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/style.module.scss:72]: added new class, removed dupe

File: components/ai_chat/resources/page/components/conversation_list/index.tsx (hunk 12 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/index.tsx]: makes sense. removed.

File: components/ai_chat/resources/page/components/conversation_list/style.module.scss (hunk 11 above)
Comment (nullhook) [reply components/ai_chat/resources/page/components/conversation_list/style.module.scss]: done
